### PR TITLE
Entity validation

### PR DESCRIPTION
**Pull request by Eyobyb**
# Your checklist for this pull request
Thank you for submitting a pull request! To speed up the review process, please follow this checklist:

- [ ] My Pull Request is small and focused on one topic so it can be reviewed easily 
- [x] My code follows the style guidelines of this project (`make format`)
- [ ] Commit messages are detailed
- [x] I have performed a self-review of my code
- [x] I commented hard-to-understand parts of my code
- [x] I updated the documentation (docstrings, `/docs`)
- [ ] My changes generate no new warnings (or explain any new warnings and why they're ok)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] All tests pass when I run `pytest tests` (offline mode)

Additional steps for code with networking dependencies:

- [ ] I followed the [offline and online testing guidelines](https://sherpa-ai.readthedocs.io/en/latest/How_To/Test/test.html#offline-and-online-testing)
- [ ] All tests pass when I run `pytest tests --external_api` (online mode, making network calls)

This is entity validation, designed to check if entities from the context exist in the answer. If they do not, the system will reprompt to add those entities. There are three stages of entity validation: first, string comparison from the list of entities extracted using spaCy; second, using metrics; and third, utilizing language model (LLM)

pervious pr #295 
issue  #274

### CODE REVIEW

**Review by oshoma — CHANGES_REQUESTED** [review]
@Eyobyb and @20001LastOrder please coordinate with each other on this Entity Validation PR (#298) and the Hydra Config PR #289. 

PR #289 contains changes to number validation and configuration settings like TEMPERATURE. Those changes need to be be picked up and carried forward in this PR as well. 

There are two options:
1. Merge Entity Validation #298 into main first, then rebase Hydra Config #298 on main (or on #298).
2. Merge Hydra Config into main first, then rebase Entity Validation on main (or on Hydra Config).

File: src/tests/unit_tests/test_util.py [hunk 1]
```diff
@@ -189,3 +194,95 @@ def test_extract_numbers_from_text_fails(source_data, incorrect_result_data):
...
+    assert result["messages"] == ""
+
+
+def test_text_similarity_with_entities_not_exist():
+    check_entity = ["apple", "orange", "banana"]
+    source_entity = ["pear", "grape", "kiwi"]
+    result = text_similarity_by_metrics(check_entity, source_entity)
+    assert result["entity_exist"] is False
+    expected_message = "remember to address these entities pear, grape, kiwi,  in final the answer."
+    print(result["messages"])
```
Comment (oshoma) [inline src/tests/unit_tests/test_util.py:285]: please remove all the print() statements throughout this PR

File: src/tests/unit_tests/test_util.py [hunk 2]
```diff
@@ -189,3 +194,95 @@ def test_extract_numbers_from_text_fails(source_data, incorrect_result_data):
...
+
+
+
+
+def test_extract_entities_with_entities():
+    text = "The United Nations (ORG) is an international organization. Some countries are members of NORP, while others are not."
+    result = extract_entities(text)
+    assert result == ["The United Nations", "NORP"]
+
+def test_extract_entities_without_entities():
```
Comment (oshoma) [inline src/tests/unit_tests/test_util.py:231]: also test extract_entities with None and a blank string ""

File: src/sherpa_ai/utils.py [hunk 3]
```diff
@@ -325,3 +330,217 @@ def check_if_number_exist(result: str, source: str):
...
+     
+    nlp = spacy.load("en_core_web_sm")
+    doc = nlp(text)
+    entity_types = ["NORP", "ORG", "GPE" , "LOC"]
+    filtered_entities = [ent.text for ent in doc.ents if ent.label_ in entity_types]
+
+    return filtered_entities
+
+
+def json_extractor(text):
```
Comment (oshoma) [inline src/sherpa_ai/utils.py:383]: json_extractor => json_from_text

File: src/sherpa_ai/utils.py [hunk 4]
```diff
@@ -325,3 +330,217 @@ def check_if_number_exist(result: str, source: str):
...
+
+    llm = SherpaOpenAI(
+        temperature=cfg.TEMPRATURE,
+        openai_api_key=cfg.OPENAI_API_KEY,
+        user_id=user_id,
+        team_id=team_id,
+    )
+
+    instruction = f"""
+        I have a question and an answer. I want you to confirm if the entities from the question has been in some form inside the answer.
```
Comment (oshoma) [inline src/sherpa_ai/utils.py:437]: I have a question and an answer. I want you to confirm whether the entities from the question are mentioned in some form within the answer.

File: src/sherpa_ai/utils.py [hunk 5]
```diff
@@ -325,3 +330,217 @@ def check_if_number_exist(result: str, source: str):
...
+    instruction = f"""
+        I have a question and an answer. I want you to confirm if the entities from the question has been in some form inside the answer. 
+        Answer = {result}
+        Question = {source}
+        Entities inside the question = {source_entity}
+       """
+    instruction = (
+        instruction
+        + """
+           only return { entity_exist: 'true' , messages:"" } if entities are mentioned inside the answer in
```
Comment (oshoma) [inline src/sherpa_ai/utils.py:445]: remove "in" from end of line 445, I think?

File: src/sherpa_ai/utils.py [hunk 6]
```diff
@@ -325,3 +330,217 @@ def check_if_number_exist(result: str, source: str):
...
+        I have a question and an answer. I want you to confirm if the entities from the question has been in some form inside the answer. 
+        Answer = {result}
+        Question = {source}
+        Entities inside the question = {source_entity}
+       """
+    instruction = (
+        instruction
+        + """
+           only return { entity_exist: 'true' , messages:"" } if entities are mentioned inside the answer in  
+           only return {entity_exist: 'false' , messages: " Entity x hasn't been mentioned inside the answer"} if the entity is not mentioned properly .
```
Comment (oshoma) [inline src/sherpa_ai/utils.py:446]: only return {entity_exist: 'false' , messages: " Entity x is not mentioned within the answer"} if the entity is not mentioned within the answer .

File: src/sherpa_ai/utils.py [hunk 7]
```diff
@@ -325,3 +330,217 @@ def check_if_number_exist(result: str, source: str):
...
+          """
+    )
+
+    llm_result = llm.predict(instruction)
+
+    checkup_json = json_extractor(llm_result)
+    return checkup_json
+
+
+def text_similarity_by_metrics(check_entity: List[str], source_entity: List[str]):
```
Comment (oshoma) [inline src/sherpa_ai/utils.py:456]: check_entity => check_entities
source_entity => source_entities

File: src/sherpa_ai/utils.py [hunk 8]
```diff
@@ -325,3 +330,217 @@ def check_if_number_exist(result: str, source: str):
...
+        dict: Result of the check containing 'entity_exist' and 'messages'.
+    """
+
+    threshold = 0.75
+    error_entity = []
+    message = ""
+    levenshtein_constant = 0.5
+    for entity in source_entity:
+        metrics_value = 0
+        for sub_entitiy in check_entity:
```
Comment (oshoma) [inline src/sherpa_ai/utils.py:474]: sub_entitiy => sub_entity

File: src/sherpa_ai/utils.py [hunk 9]
```diff
@@ -325,3 +330,217 @@ def check_if_number_exist(result: str, source: str):
...
     return {"number_exists": True, "messages": message}
+
+def string_comparison_with_jaccard_and_levenshtein(word1, word2, levenshtein_constant):
+    """
+        Calculate a combined similarity metric using Jaccard similarity and normalized Levenshtein distance.
+
+        Args:
+        - word1 (str): First input string.
+        - word2 (str): Second input string.
+        - levenshtein_constant (float): Weight for the Levenshtein distance in the combined metric.
```
Comment (oshoma) [inline src/sherpa_ai/utils.py:341]: levenshtein_constant  => levenshtein_weight 

levenshtein_weight (float):  Weight between 0 and 1.0 for the Levenshtein distance in the combined metric.

File: src/sherpa_ai/utils.py [hunk 10]
```diff
@@ -325,3 +330,217 @@ def check_if_number_exist(result: str, source: str):
...
+    for entity in source_entity:
+        metrics_value = 0
+        for sub_entitiy in check_entity:
+            word1 = entity
+            word2 = sub_entitiy
+
+            combined_similarity = string_comparison_with_jaccard_and_levenshtein(
+                word1=word1, word2=word2, levenshtein_constant=levenshtein_constant
+            )
+            if metrics_value < combined_similarity:
```
Comment (oshoma) [inline src/sherpa_ai/utils.py:481]: metrics_value = max(combined_similarity, metrics_value)

File: src/sherpa_ai/utils.py [hunk 11]
```diff
@@ -325,3 +330,217 @@ def check_if_number_exist(result: str, source: str):
...
+
+        Args:
+        - check_entity (List[str]): List of entities to check.
+        - source_entity (List[str]): List of reference entities.
+
+        Returns:
+        dict: Result of the check containing 'entity_exist' and 'messages'.
+    """
+
+    threshold = 0.75
```
Comment (oshoma) [inline src/sherpa_ai/utils.py:468]: threshold => probably_mentioned

File: src/sherpa_ai/utils.py [hunk 12]
```diff
@@ -325,3 +330,217 @@ def check_if_number_exist(result: str, source: str):
...
+            combined_similarity = string_comparison_with_jaccard_and_levenshtein(
+                word1=word1, word2=word2, levenshtein_constant=levenshtein_constant
+            )
+            if metrics_value < combined_similarity:
+                metrics_value = combined_similarity
+        if metrics_value < threshold:
+            error_entity.append(entity)
+
+    if len(error_entity) > 0:
+        for entity in error_entity:
```
Comment (oshoma) [inline src/sherpa_ai/utils.py:487]: entities = ", ".join(error_entity)
message = f"remember to address these entities {entities} in final the answer."

File: src/sherpa_ai/utils.py [hunk 13]
```diff
@@ -325,3 +330,217 @@ def check_if_number_exist(result: str, source: str):
...
+            message += entity + ", "
+        message = f"remember to address these entities {message} in final the answer."
+        return {"entity_exist": False, "messages": message}
+    return {"entity_exist": True, "messages": message}
+
+
+def text_similarity(check_entity: List[str], source_entity: List[str]):
```
Comment (oshoma) [inline src/sherpa_ai/utils.py:494]: check_entity => check_entities
source_entity => source_entities

File: src/sherpa_ai/utils.py [hunk 14]
```diff
@@ -325,3 +330,217 @@ def check_if_number_exist(result: str, source: str):
...
+        - check_entity ([str]): List of entities to check.
+        - source_entity ([str]): List of reference entities.
+
+        Returns:
+        dict: Result of the check containing 'entity_exist' and 'messages'.
+    """
+
+    error_entity = []
+    message = ""
+    for entity in source_entity:
```
Comment (oshoma) [inline src/sherpa_ai/utils.py:508]: use Python set difference instead of a for loop

File: src/sherpa_ai/utils.py [hunk 15]
```diff
@@ -325,3 +330,217 @@ def check_if_number_exist(result: str, source: str):
...
+        print(entity)
+        if entity not in check_entity:
+            error_entity.append(entity)
+    if len(error_entity) > 0:
+        for entity in error_entity:
+            message += entity + ", "
```
Comment (oshoma) [inline src/sherpa_ai/utils.py:514]: message += ", ".join(error_entity)

File: src/sherpa_ai/utils.py [hunk 16]
```diff
@@ -325,3 +330,217 @@ def check_if_number_exist(result: str, source: str):
...
+        - result (str): Answer text.
+        - source (str): Question text.
+        - stage (int): Stage of the check (0, 1, or 2).
+
+        Returns:
+        dict: Result of the check containing
+    """
+
+    source_entity = extract_entities(source)
+    check_entity = extract_entities(result)
```
Comment (oshoma) [inline src/sherpa_ai/utils.py:534]: Let's rework this `stage` variable. You are implementing a state machine here. So give the states meaningful names, e.g. TEXT_SIMILARITY_BASIC = 0, TEXT_SIMILARITY_BY_METRICS = 1, TEXT_SIMILARITY_BY_LLM = 2. Something like that. 

If function callers must know what these states mean then define them at the top of this file so they are obvious. 

Alternatively, is there a way to hide all of this from function callers? It seems like implementation detail to me, but I haven't been part of the design conversation.

File: src/tests/integration_tests/test_entity_citation_validator.py [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+from sherpa_ai.models.sherpa_base_chat_model import SherpaChatOpenAI
+
+# from sherpa_ai.test_utils.llms import get_llm
+from sherpa_ai.tools import SearchTool
+import sherpa_ai.config as cfg
+from unittest.mock import patch
+from sherpa_ai.utils import extract_entities
+from nltk.metrics import jaccard_distance
+
+@pytest.mark.parametrize(
```
Comment (oshoma) [inline src/tests/integration_tests/test_entity_citation_validator.py:17]: Hooray for parametrize! 🥇

File: src/tests/integration_tests/test_entity_citation_validator.py [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+        )
+
+        task_agent.run()
+
+        results = shared_memory.get_by_type(EventType.result)
+        logger.error(results[0].content)
+        result_entities = extract_entities(results[0].content)
+        for entitity in expected_entities:
+            set_a = set(entitity.split())  # Convert each string in a to a set of words
+            match_found = any( 1 - jaccard_distance(set_a, result_entity.split()) >= 0.7 for result_entity in result_entities)
```
Comment (oshoma) [inline src/tests/integration_tests/test_entity_citation_validator.py:81]: here you are using 0.7, and elsewhere in the code there is a 0.75 value. are they meant to be the same? if so, define a constant and use it (import it) in the test.

File: src/tests/unit_tests/test_util.py [hunk 19]
```diff
@@ -189,3 +194,95 @@ def test_extract_numbers_from_text_fails(source_data, incorrect_result_data):
     # test aganist a text which don't have the same numers as the source
     check_result = check_if_number_exist(incorrect_result_data, source_data)
     assert not check_result["number_exists"]
+
+def test_json_extractor_valid_json():
```
Comment (oshoma) [inline src/tests/unit_tests/test_util.py:198]: nice tests.
these tests could also be made more concise with parametrize.

remember to check the failure cases and corner cases like `None` and `""`

**Review by oshoma — CHANGES_REQUESTED** [review]

File: src/sherpa_ai/utils.py [hunk 20]
```diff
@@ -325,3 +330,217 @@ def check_if_number_exist(result: str, source: str):
...
+            error_entity.append(entity)
+    if len(error_entity) > 0:
+        for entity in error_entity:
+            message += entity + ", "
+        message = f"remember to address these entities {message} in final the answer."
+        return {"entity_exist": False, "messages": message}
+    return {"entity_exist": True, "messages": message}
+
+
+def check_entities_match(result: str, source: str, stage: int):
```
[the remark this answers is not in the document]
Comment (Eyobyb) [reply src/sherpa_ai/utils.py:520]: we talked about it and decided to move this to the Hydra Config after merging it

**Comment by oshoma** [comment]
> @Eyobyb and @20001LastOrder please coordinate with each other on this Entity Validation PR (#298) and the Hydra Config PR #289. PR #289 contains changes to number validation and configuration settings like TEMPERATURE. Those changes need to be be picked up and carried forward in this PR as well.
> 
> There are two options:
> 
> 1. Merge Entity Validation [Entity validation #298](https://github.com/Aggregate-Intellect/sherpa/pull/298) into main first, then rebase Hydra Config [Entity validation #298](https://github.com/Aggregate-Intellect/sherpa/pull/298) on main (or on [Entity validation #298](https://github.com/Aggregate-Intellect/sherpa/pull/298)).
> 2. Merge Hydra Config into main first, then rebase Entity Validation on main (or on Hydra Config).

Noting that Hydra Config PR #289 is now merged into main. @Eyobyb please rebase your Entity Validation work on main.
